### PR TITLE
feat: Helm charts for deployment

### PR DESCRIPTION
**Pull request by jlehtimaki**

### CODE REVIEW

File: charts/endpoint-controller/templates/clusterrole.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: endpoint-controller
```
Comment (kayano) [inline charts/endpoint-controller/templates/clusterrole.yaml:4]: I guess we should use template `endpoint-controller.fullname` from helpers

File: charts/endpoint-controller/templates/clusterrole.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  resources: ["services"]
+  verbs: ["get", "list", "watch"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: endpoint-controller
+subjects:
+- kind: ServiceAccount
+  name: endpoint-controller
```
Comment (kayano) [inline charts/endpoint-controller/templates/clusterrole.yaml:19]: shouldn't be hardcoded

File: charts/endpoint-controller/templates/clusterrole.yaml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  resources: ["endpoints"]
+  verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
+- apiGroups: [""]
+  resources: ["services"]
+  verbs: ["get", "list", "watch"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: endpoint-controller
```
Comment (kayano) [inline charts/endpoint-controller/templates/clusterrole.yaml:16]: I guess we should use template endpoint-controller.fullname from helpers for a name here

File: charts/endpoint-controller/templates/clusterrole.yaml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+subjects:
+- kind: ServiceAccount
+  name: endpoint-controller
+  namespace: default
+roleRef:
+  kind: ClusterRole
+  name: endpoint-controller
```
Comment (kayano) [inline charts/endpoint-controller/templates/clusterrole.yaml:23]: shouldn't be harcoded

File: charts/endpoint-controller/templates/deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,55 @@
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ .Values.name }}
```
Comment (kayano) [inline charts/endpoint-controller/templates/deployment.yaml:5]: usually it is taken from template `endpoint-controller.fullname`

File: charts/endpoint-controller/templates/deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    spec:
+      serviceAccountName: {{include "endpoint-controller.serviceAccountName" .}}
+      containers:
+        - name: {{ .Chart.Name }}
+          image: "{{.Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion}}"
+          imagePullPolicy: {{.Values.image.pullPolicy}}
+          environment:
+              SYNC_PERIOD: {{.Values.controller.sync_period}}
+              BLOCK_MISS: {{.Values.controller.block_miss}}
+          securityContext:
```
Comment (kayano) [inline charts/endpoint-controller/templates/deployment.yaml:30]: Don't we want to allow to override securityContext  and take default from values.yaml?

File: charts/endpoint-controller/templates/deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+      labels:
+        {{- include "endpoint-controller.selectorLabels" . | nindent 8 }}
+    spec:
+      serviceAccountName: {{include "endpoint-controller.serviceAccountName" .}}
+      containers:
+        - name: {{ .Chart.Name }}
+          image: "{{.Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion}}"
+          imagePullPolicy: {{.Values.image.pullPolicy}}
+          environment:
+              SYNC_PERIOD: {{.Values.controller.sync_period}}
```
Comment (kayano) [inline charts/endpoint-controller/templates/deployment.yaml:28]: We have sane defaults in controller, I would rather allow to provide extra env vars from values.yaml

File: charts/endpoint-controller/templates/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  template:
+    metadata:
+      {{- with .Values.podAnnotations }}
+      annotations:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      labels:
+        {{- include "endpoint-controller.selectorLabels" . | nindent 8 }}
+    spec:
+      serviceAccountName: {{include "endpoint-controller.serviceAccountName" .}}
```
Comment (kayano) [inline charts/endpoint-controller/templates/deployment.yaml:22]: Should we check `.Values.serviceAccount.create` before we render it?

File: charts/endpoint-controller/templates/deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,55 @@
+---
+apiVersion: apps/v1
+kind: Deployment
```
Comment (kayano) [inline charts/endpoint-controller/templates/deployment.yaml:3]: Would be good to add some healthchecks for endpoint-controller and define probes for the deployment

File: charts/endpoint-controller/templates/deployment.yaml (hunk 7 above)
Comment (jlehtimaki) [reply charts/endpoint-controller/templates/deployment.yaml:28]: for what purpose?
I think it's easier for end user to have it like this.

File: charts/endpoint-controller/templates/deployment.yaml (hunk 6 above)
Comment (jlehtimaki) [reply charts/endpoint-controller/templates/deployment.yaml:32]: sure

File: charts/endpoint-controller/templates/deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+      {{- end }}
+      labels:
+        {{- include "endpoint-controller.selectorLabels" . | nindent 8 }}
+    spec:
+      serviceAccountName: {{include "endpoint-controller.serviceAccountName" .}}
+      containers:
+        - name: {{ .Chart.Name }}
+          image: "{{.Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion}}"
+          imagePullPolicy: {{.Values.image.pullPolicy}}
+          env:
```
Comment (kayano) [inline charts/endpoint-controller/templates/deployment.yaml:28]:
```
Error: Failed to render chart: exit status 1: Error: unable to build kubernetes objects from release manifest: error validating "": error validating data: [ValidationError(Deployment.spec.template.spec.containers[0].env): invalid type for io.k8s.api.core.v1.Container.env: got "map", expected "array", ValidationError(Deployment.spec.template.spec.containers[0].securityContext): unknown field "fsGroup" in io.k8s.api.core.v1.SecurityContext]
```

**Review by kayano — APPROVED** [review]

**Pull request merged**